node/src/components/diagnostics_port/tasks.rs

```diff
@@ -0,0 +1,630 @@
+            File,
+            WithOtherTrailing<WithOtherIntEncoding<DefaultOptions, FixintEncoding>, AllowTrailing>,
+        >,
+    ),
+}
+
+impl FileSerializer {
+    /// Converts the temp file serializer into an actual erased serializer.
+    pub fn as_serializer<'a>(&'a mut self) -> Box<dyn ErasedSerializer + 'a> {
+        match self {
+            FileSerializer::Json(json) => Box::new(<dyn erased_serde::Serializer>::erase(json)),
+            FileSerializer::Bincode(bincode) => {
+                Box::new(<dyn erased_serde::Serializer>::erase(bincode))
+            }
+        }
+    }
+}
+
+/// Error obtaining a queue dump.
+#[derive(Debug, Error)]
+enum ObtainDumpError {
+    /// Error trying to create a temporary directory.
+    #[error("could not create temporary directory")]
+    CreateTempDir(#[source] io::Error),
+    /// Error trying to create a file in the temporary directory.
+    #[error("could not create file in temporary directory")]
+    CreateTempFile(#[source] io::Error),
+    /// Error trying to reopen the file in the temporary directory after writing.
+    #[error("could not reopen file in temporary directory")]
+    ReopenTempFile(#[source] io::Error),
+}
+
+impl Session {
+    /// Creates a serializer for an `EraDump`.
+    fn create_era_dump_serializer(&self) -> fn(&EraDump<'_>) -> Result<Vec<u8>, Cow<'static, str>> {
+        // TODO: This function could probably be a generic serialization function for any `T`, but
+        // the conversion is tricky due to the lifetime arguments on `EraDump` and has not been done
+        // yet.
+        match self.output {
+            OutputFormat::Interactive => |data: &EraDump| {
+                let mut buf = data.to_string().into_bytes();
+                buf.push(b'\n');
+                Ok(buf)
+            },
+            OutputFormat::Json => |data: &EraDump| {
+                let mut buf = serde_json::to_vec(&data).map_err(|err| {
+                    Cow::Owned(format!("failed to serialize era dump as JSON: {}", err))
+                })?;
+                buf.push(b'\n');
+                Ok(buf)
+            },
+            OutputFormat::Bincode => |data: &EraDump| {
+                bincode::serialize(&data).map_err(|err| {
+                    Cow::Owned(format!("failed to serialize era dump as bincode: {}", err))
+                })
+            },
+        }
+    }
+
+    /// Creates a generic serializer that is writing to a temporary file.
+    ///
+    /// The resulting serializer will write to the given file.
+    fn create_queue_dump_format(&self, file: File) -> QueueDumpFormat {
+        match self.output {
+            OutputFormat::Interactive => QueueDumpFormat::debug(file),
+            OutputFormat::Json => {
+                QueueDumpFormat::serde(FileSerializer::Json(serde_json::Serializer::new(file)))
+            }
+            OutputFormat::Bincode => {
+                QueueDumpFormat::serde(FileSerializer::Bincode(bincode::Serializer::new(
```

Comment:
I wonder, whether we would ever want to use `bincode` - it's not very human readable.

Reply:
For large dumps that we have special clients for.

We could drop it though, that's a good point; after all, we can on-the-fly gzip larger ones. :thinking:

Reply:
>For large dumps that we have special clients for.

We have clients for that? Who?

Reply:
Err, might write, sorry :)